src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -390,6 +390,54 @@ private static bool CheckConstraintsSingleType(TypeSymbol type, CheckConstraints
         }
 
         public static bool CheckConstraints(
+            this TupleTypeSymbol tuple,
+            ConversionsBase conversions,
+            SyntaxNode typeSyntax,
+            ImmutableArray<Location> elementLocations,
+            Compilation currentCompilation,
+            DiagnosticBag diagnostics)
+        {
+            var type = tuple.TupleUnderlyingType;
+
+            ArrayBuilder<TypeParameterDiagnosticInfo> useSiteDiagnosticsBuilder = null;
+
+            var result = !typeSyntax.HasErrors;
+
+            var diagnosticsBuilder = ArrayBuilder<TypeParameterDiagnosticInfo>.GetInstance();
+
+            var underlyingTupleTypeChain = ArrayBuilder<NamedTypeSymbol>.GetInstance();
+            TupleTypeSymbol.GetUnderlyingTypeChain(type, underlyingTupleTypeChain);
+
+            var offset = 0;
+            foreach (var underlyingTuple in underlyingTupleTypeChain)
+            {
+                result &= CheckTypeConstraints(underlyingTuple, conversions, currentCompilation, diagnosticsBuilder, ref useSiteDiagnosticsBuilder);
+
+                if (useSiteDiagnosticsBuilder != null)
+                {
+                    diagnosticsBuilder.AddRange(useSiteDiagnosticsBuilder);
+                }
+
+                foreach (var pair in diagnosticsBuilder)
+                {
+                    var ordinal = pair.TypeParameter.Ordinal + offset;
+                    var location = ordinal < elementLocations.Length ? elementLocations[ordinal] : typeSyntax.Location;
+                    diagnostics.Add(new CSDiagnostic(pair.DiagnosticInfo, location));
+                }
+
+                useSiteDiagnosticsBuilder = null;
+                diagnosticsBuilder.Clear();
+
+                offset += 7;
```

Comment:
@jcouv Is that ok to introduce `internal const int RestIndex = RestPosition - 1;` ?